tools/outline_proxy_controller/outline_proxy_controller.cpp

```diff
@@ -0,0 +1,587 @@
+
+  if (!isSuccessful(tunDeviceAdditionResult)) {
+    logger.error(tunDeviceAdditionResult.first);
+    throw runtime_error("failed to set the tun device ip address");
+  }
+}
+
+void OutlineProxyController::detectBestInterfaceIndex() {
+  SubCommand getRouteCommand;
+
+  // our best guest is the route that outline server already can be reached
+  // it is the default gateway if outline is connected or not
+  getRouteCommand.push(make_pair("get", outlineServerIP));
+  auto result = executeIPRoute(getRouteCommand);
+
+  if (!isSuccessful(result)) {
+    logger.error(result.first);
+    throw runtime_error("unable to query the default route to the outline proxy");
+  }
+
+  std::string routingData = result.first;
+
+  try {
+    routingGatewayIP = getParamValueInResult(routingData, "via");
+    clientToServerRoutingInterface = getParamValueInResult(routingData, "dev");
+    clientLocalIP = getParamValueInResult(routingData, "src");
+  } catch (runtime_error& e) {
+    logger.error(e.what());
+    throw runtime_error("Failed to parse the routing query response");
+  }
+}
+
+void OutlineProxyController::routeThroughOutline(std::string outlineServerIP) {
+  // Sanity checks
+  if (outlineServerIP.empty()) {
+    logger.error("attempt to route throw a server whose IP we are not informed of");
+    throw runtime_error("outlineServerIP is empty");
+  }
+
+  logger.info("attempting to route through outline server " + outlineServerIP);
+
+  // TODO: make sure the routing rule isn't already in the table
+  if (routingStatus == ROUTING_THROUGH_OUTLINE) {
+    logger.warn("it seems that we are already routing through outline server");
+  }
+
+  this->outlineServerIP = outlineServerIP;
+
+  backupDNSSetting();
+
+  try {
+    createRouteforOutlineServer();
+  } catch (exception& e) {
+    // we can not continue
+    logger.error("failed to create a proirity route to outline proxy: " + string(e.what()));
+    // We failed to make a route through outline proxy. We just remove the flag
+    // indicating DNS is backed up.
+    resetFailRoutingAttempt(OUTLINE_PRIORITY_SET_UP);
+    return;
+  }
+
+  try {
+    deleteAllDefaultRoutes();  // drop the default route before adding another one
+  } catch (exception& e) {
+    logger.error("failed to remove the default route throw the current default router: " +
+                 string(e.what()));
+    resetFailRoutingAttempt(DEFAULT_GATEWAY_ROUTE_DELETED);
+    return;
+  }
+
+  try {
+    createDefaultRouteThroughTun();
+  } catch (exception& e) {
+    logger.error("failed to route network traffic through outline tun interfacet: ", e.what());
+    resetFailRoutingAttempt(TRAFFIC_ROUTED_THROUGH_TUN);
+    return;
+  }
+
+  try {
+    toggleIPv6(false);
+  } catch (exception& e) {
+    // We are going ahead with routing through Outline even if we are not able
+    // to disable all IPV6 routes.
+    logger.warn("possible net traffic leakage. failed to disable IPv6 routes on all interfaces: " +
```

Comment:
Let's fail, since we fail on all other "possible leakage" events.

Reply:
I changed the warning message to error. However, I am not sure we should fail altogether if IPv6 is possibly leaking. Let me know, if it's necessary.

Reply:
Yes, please, let's fail if IPv6 is leaking. This is what we do on the other platforms.

Reply:
Done.